### ISSUE TITLE
the request is for a topic or partition that does not exist on this broker

### ISSUE THREAD

**Issue by brzyangg** [segmentio/kafka-go#516]
**Describe the bug**

[3] Unknown Topic Or Partition: the request is for a topic or partition that does not exist on this broker

but my topic is exist

**Kafka Version**

v2.2.1

**To Reproduce**

Occasionally when writing a message

I am new a writer and write a message

**Expected behavior**

I hope to solve this problem

**Additional context**

broker number : 3
partition number: 3
replicas: 3

**Comment by collinvandyck** [comment]
Hi @brzyangg ! This happens sometimes with brand new topics, as there is a delay between when the topic is created on the broker, and when the topic can actually be written to. If writes happen during this time period, this error may occur. Does this describe what's happening to you?

**Comment by achille-roussel** [comment]
@brzyangg would you be able to provide the version of kafka-go that you are using?

Do you happen to have a minimal program that exhibits the problem that you could share with us?

**Comment by brzyangg** [comment]
@achille-roussel My kafka-go version: v0.3.7
my minimal program:
```go
config := kafka.WriterConfig{
	Brokers:    brokers,
	Topic:        topic,
	Balancer:  &kafka.Hash{},
	BatchSize: 1,
}
writer := kafka.NewWriter(config)
writer.WriteMessages(ctx, kafka.Message{
	Key:   []byte(k),  // k is id 
	Value: msg,
})
```

**Comment by achille-roussel** [comment]
Are you able to reproduce consistently with this configuration?

**Comment by brzyangg** [comment]
No, this problem is not obligatory, it will appear several times a day

**Comment by oneslash** [comment]
On AWS MSK, we don't have a problem with v0.3.10 however started having this issue on v0.4.x.
Apache MSK version 2.5.1

**Comment by filinvadim** [comment]
So?

**Comment by brzyangg** [comment]
I have solved this problem, thank you all.

Is the cause of my misconfiguration

**Comment by matzew** [comment]
@brzyangg mind explaining what was the fix for this ?

**Comment by brzyangg** [comment]
@matzew Topic we modified in a hidden place.

**Comment by hieudoan7** [comment]
can you elaborate more on what is 'hidden place'?

**Comment by brzyangg** [comment]
sorry, excuse everyone. This `hidden place` refers to the internal program, not kafka-go. This problem has been solved.

**Comment by ekeric13** [comment]
So I sporadically get this issue as well...
The topic my client is trying to write to is weeks old... so definitely not a new topic issue.

I am using kafka 2.6.2 via msk.

I am actually using msk sasl IAM authentication when communicating to the broker, not sure if that matters.
My setup looks like this:

```
	sharedTransport := &kafka.Transport{}
	if useMskIam {
		awsCreds := credentials.NewEnvCredentials()

		sharedTransport = &kafka.Transport{
			TLS: &tls.Config{},
			SASL: &aws_msk_iam.Mechanism{
				Signer: sigv4.NewSigner(awsCreds),
				Region: awsRegion,
			},
		}
	}

	kafkaService := &kafka.Writer{
		Addr:  kafka.TCP(strings.Split(brokersEndpoint, ",")...),
		Topic: kafkaTopic,
		// default hash algo used by kafka
		Balancer:     kafka.Murmur2Balancer{},
		RequiredAcks: kafka.RequireOne,
		Transport:    sharedTransport,
	}

```

And then the part of the code where I try to write to kafka and get the error looks like this:
```
    err = k.kafkaService.WriteMessages(context.Background(),
		kafka.Message{
			Key:     []byte(*event.Key),
			Value:   []byte(*event.Value),
			Headers: headers,
		},
	)
// err.Error() == "[3] Unknown Topic Or Partition: the request is for a topic or partition that does not exist on this broker"
```

**Comment by ekeric13** [comment]
Okay I think my issue may have been my MSK IAM permissions being too strict.

**Comment by jacksenk** [comment]
I recently had a similar problem and believe I've found the cause. We have a bunch of servers writing to Kafka 24x7, and we regularly every ~12 hours get the same error: "[3] Unknown Topic Or Partition: the request is for a topic or partition that does not exist on this broker" which would resolve itself after a couple of minutes

We're using MSK with IAM authentication. What we discovered is the brokers have a config option listener.name.client_iam_public.aws_msk_iam.connections.max.reauth.ms which is set to 12 hours. I believe the credentials are expiring, resulting in these errors, the client eventually reconnects after enough failed retries which results in a new session and everything works again until the next ~12 hours where the cycle repeats
